### PR TITLE
feat(docs): add links to runnable examples in sequential-model-guide.md

### PR DESCRIPTION
**Pull request by CarmenPopoviciu**
This commit adds links to runnable [MachineLabs](https://machinelabs.ai) executions for all examples on the ["Guide to the Sequential model"](https://keras.io/getting-started/sequential-model-guide) documentation page.

Each complete example in the documentation will now have a dedicated "lab" in MachineLabs, with its link pointing to that lab's original execution (see picture below)

screen shot 2017-08-23 at 22 16 50

### CODE REVIEW

**Comment by fchollet** [comment]
A few comments:

- MachineLab does not appear to be able to properly render the Keras progbar. This is a problem. Ideally it should behave like a local terminal.
- The link should lead to the editor view rather than the console view.
- It's impossible to run code without logging in. If we are sending users to an external location with a promise that they will be able to run the example they just saw, then they should be able to run it, otherwise we are tricking them. An honest link would then read "click here to sign up to MachineLab and run this example", which is accurate, but is basically an ad. Most JS services that offer similar functionality don't require signing up in order to run code. You'll also find out that forcing users to sign up for something in order to use it in a way that doesn't conceptually require an account, does not lead to *actual* conversions, only to a bump in vanity metrics. You want happy and engaged users, not just a counter that goes up.
- On a similar note, the links shouldn't read "Try it on MachineLab", because it sounds like an ad. For instance, on the jQuery website, the live example pages (e.g. `http://try.jquery.com/`) only make a footnote mention of the provider. Ultimately, readers of the Keras docs don't care about trying something *on MachineLab*, they care about trying in immediately in their browser -- so that's what the links should convey, e.g. `click here to try it in your browser`. The docs should serve the users; they should not serve as a conduit for external marketing.

**Comment by cburgdorf** [comment]
Christoph here from MachineLabs. First off, thank you @CarmenPopoviciu for creating the PR!

And thank you @fchollet for your valuable feedback! Let me jump right into yourn points.

>MachineLab does not appear to be able to properly render the Keras progbar

Are we talking about those two lines on the screenshot that are shifted to the right? We'll look into it.

>The link should lead to the editor view rather than the console view

We can add a URL param to control which view should open.

>It's impossible to run code without logging in.

That's true. However, you can view the output and the code without logging in. The reason you have to login to execute the code is that we have to protect ourselves from malicious users. If we would open up the system so that every anonymous user could run code on our servers we would be open to DOS attacks among other attacks. We need a way to identify the user so that we can track how much computation time they have already spend.

Not only that but also giving anonymous user access to backend code with internet access would put us at risk of people using our service to do illegal stuff without us being able to track it.

>Most JS services that offer similar functionality don't require signing up in order to run code

That is true, but the difference is that there code runs *in the browser of the user* which means there's nothing a malicious user could attack except their own computer when it comes to consuming resources.

So, as much as we would love to enable every anonymous user to just instantly run the code, I don't see a feasible way to do that. That said, all you have to do with MachineLabs is really just to login (and we have login with Google in the making) and once you are logged in, every user will have some monthly free CPU hours to burn for ever. Probably something like 75 free hours per month.

>On a similar note, the links shouldn't read "Try it on MachineLab" 

We were inspired by the React docs that read ["Try it on CodePen"](https://facebook.github.io/react/docs/state-and-lifecycle.html). But if your would rather like us to change this to "Try it in your browser" that would be cool with us, too. Our main goal really is to make it easier for the user to get their feet wet with Machine Learning code.

**Comment by CarmenPopoviciu** [comment]
Thank you François for the feedback and Chris for clarifying the questions! 

I am very happy to make the requested changes. I think linking to the editor view and changing the link text is very straightforward to do. The progbar issue is something that Christof will probably look into. That leaves us with just the issue of logging in. For now I'd rather wait for François' feedback on that, to be sure that is not a game stopper on his side

**Comment by cburgdorf** [comment]
> Ideally it should behave like a local terminal

Oh, just tried it out locally and noticed the difference. We've had it already on our roadmap to make our console behave more like an actual terminal. We can shift priorities to fix that earlier than what we had planned. In fact, we just started to look into it and may have something ready sooner than what we initially thought :)

**Comment by fchollet** [comment]
> That leaves us with just the issue of logging in. For now I'd rather wait for François' feedback on that, to be sure that is not a game stopper on his side

I don't really understand the DOS mitigation argument. A GitHub login or email login does not serve as a unique human user identifier any more than an IP address. If I can DOS you when your site does not require logging in, then I can DOS you when your site requires a GitHub or email login, and the amount of additional effort on my part is negligible.

In fact, it is *easier* to DOS your site if your only DOS mitigation measure is a login, because new logins are far easier to obtain than new IPs. New IPs are always the bottleneck for people trying to by-pass rate limiting measures.

You should do IP-based rate limiting. Plus, if you suspect a bot (e.g. based on browser signature), ask to solve a Captcha. This is the definitive way to achieve unique human user recognition.

Overall, there is no real technical argument for forcing users to login, while the damage to UX is huge. I'm saying this for you, for the sake of the success of your app.

**Comment by cburgdorf** [comment]
>Overall, there is no real technical argument for forcing users to login, while the damage to UX is huge. I'm saying this for you, for the sake of the success of your app.

I do appreciate the discussion and we've been discussing this internally a lot, too.

Maybe let's recap the power that we currently grant to logged in users:

- access ready to use docker containers with TensorFlow, Keras, Python etc pre-installed
- can run any code without restrictive limits on the duration (code can run for several days)
- they get full access to the internet.

However, we do track the number of parallel processes and limit it to two parallel processes per user and also limit the overall free CPU processing time to 72 hours per month. This is btw, way less restrictive than what [floydhub](https://www.floydhub.com/pricing) offers in their free plan. Which is only 1 concurrent job, 20 free CPU hours per month and a maximum job time of 6 hours.

Notice that if we would grant anonymous users the right to start executions we would basically lose control over these limits. You can't restrict the monthly number of free hours or concurrent jobs if people can keep launching those anonymously. We could couple those restrictions to an IP so that effectively IP w.x.y.z becomes restricted to only use 72 hours of free CPU time and have a maximum of 2 concurrent executions but that would penalize multiple anonymous user who are all sitting behind one IP (like companies, universities etc).

But that's just one side of it.

With full access to the internet and the ability to execute potentially long running code on our servers people can certainly do a lot of harm. Technically there is nothing forcing you to use MachineLabs for Machine Learning stuff. You can just go and build a spam server instead :)

Of course people can still do these harmful things even with the account model but at least we get to know which user performed some illegal task and can block their account. Again of course we could block an IP as well but I wonder how effective that would be if people can just walk into any random café place to launch their malicious code from there.

Honestly, we really want to lower the barrier of entry as much as possible but I'd be too scared to give such powers into the hand of anonymous users. Do you have any example for another service that lets anonymous users run user provided backend code with internet access? I can't imagine Google or any other company giving anonymous users a 24 hour free trial of a VPS server with full internet access for instance? 

The closet thing I can think of are sites like [play.rust-lang.org](http://play.rust-lang.org) that lets you fiddle with a language. Those usually run in well defined boundaries [without internet access](https://github.com/rust-lang/rust-playpen/blob/7c04bc3a48c2273c78428b17daa7a3be52ab319f/src/docker.rs#L24) and [a maximum execution time of 5 seconds](https://github.com/rust-lang/rust-playpen/blob/7c04bc3a48c2273c78428b17daa7a3be52ab319f/src/lib.rs#L116)

I believe that if we we would grant anonymous users the right to launch executions then these executions would also have to have a very limited execution time (just enough to see if the code works. Rather seconds than minutes) and also should not have the right to access the internet or just limited access to white-listed sites ton consume public data sets.

But that would certainly degrade the user experience a lot.

**Comment by fchollet** [comment]
> that would penalize multiple anonymous user who are all sitting behind one IP

This is not an issue if you do account-based rate limiting too and have a clear error message to tell users when their current IP is being shared and has exceeded its rate limit, asking users to create an account. You could also use browser signatures to distinguish between different legitimate users of the same IP.

My argument is really very simple: everything you say is true, but it is just as true when doing account-based rate limiting as when doing IP-based rate limiting. In fact, for an adversary, account-based is easier to by-pass than IP-based (assuming that you are not doing both).

Thus forcing users to be logged in does not in fact buy you any added protection, while it does break the UX. A random email or random Github account is not less anonymous than an IP. The few identifiers that would be better than IPs would be FB accounts or verified email addresses from private domains (e.g. a whitelist of university domains).

> With full access to the internet and the ability to execute potentially long running code on our servers people can certainly do a lot of harm

Yes, and forcing users to log in does not change any of that. Again, it is harder to obtain new IPs than to obtain new accounts, and account-based identification does not buy anything over IP-based identification, because email addresses are trivial to obtain and adversaries can be assumed to have accessed to hundreds of thousands of them.

**Comment by cburgdorf** [comment]
>This is not an issue if you do account-based rate limiting too and have a clear error message to tell users when their current IP is being shared and has exceeded its rate limit, asking users to create an account. You could also use browser signatures to distinguish between different legitimate users of the same IP.

Yes, you can probably make this work if you do IP based limitations in addition to account based limitations and ask user to login when they hit IP based limits. However, it adds a lot of complexity to our system. Especially as a small team we have to decide how we prioritize things because their is a lot of stuff we want to get done to assist with the actual ML tasks.

>Again, it is harder to obtain new IPs than to obtain new accounts

Depends. For instance we rely on Firebase for all the auth stuff and they already do an IP based rate limit for creating user accounts so that it prevents you from generating lots of accounts from one IP in a small amount of time you.

I would also expect that Google has systems in place to make it hard to mass create email addresses. 

However, I do appreciate the discussion and I see that you are giving your advice to help us make a better product. We discussed this internally again and as we are very very commit to lower the barrier of entry to ML for everyone here is what I think we could put on the roadmap.

1. Allow anonymous users not just to view (as of today) but also to start executions but limit the maximum run time to something that is enough to see that the code works but is infeasible for actual training.

2. Restrict internet access and quota for anonymous users (IP based quotas then). Not sure yet how exactly but maybe with whitelisted domains to access datasets from.

We should put that on our road map but I hope that it won't be a blocker for including those links in the documentation. We are currently hard at work with things like allowing executions to save files (trained model, generated graphics etc) and make those consumable in UI and via REST API.

This will open up a lot of cool use cases. Things like creating a jsFiddle that uses keras.js and consumes a trained model via REST API. So you can basically cover the entire story of Keras training plus web app without leaving your browser.

Meanwhile we have fixed our output. We integrated xterm.js (used by Visual Studio Code and many others) so that the output pane behaves like a real TTY.

**Comment by cburgdorf** [comment]
>Meanwhile we have fixed our output. We integrated xterm.js (used by Visual Studio Code and many others) so that the output pane behaves like a real TTY

I realized that the existing executions in this PR have some minor glitches as they were generated before we introduced the terminal modus. We applied some fixes to make the new terminal modus work with executions that were generated before but it seems there are still some minor bugs (like missing new lines) and that the easiest way to fix them would be to just re-run the executions, adjust labels and then update this PR with the new links accordingly.

@CarmenPopoviciu would you mind updating the PR

**Comment by CarmenPopoviciu** [comment]
@cburgdorf sure, I'll update the PR

**Comment by CarmenPopoviciu** [comment]
@fchollet as per your comments, the PR is now updated

**Comment by fchollet** [comment]
Cool, sounds good!

@cburgdorf the lowered access restrictions you propose would certainly improve the user experience! Basically my threshold I that I should be able to make a simple change on any example (e.g. add or remove a layer), click "run", and see results. For the time being, this is still not possible, as clicking "run" without being logged in triggers the dialog:

`You're just one click away from running that code! No complicated setup, no credit cards! All you need to do is to login with your GitHub account.`

When it becomes possible to edit and run without logging in, I will merge this PR. Restricting quota 
and runtime for anonymous users does not sound like an issue. Restricting internet access is likely an issue since many of our examples use datasets (like MNIST) that get downloaded on the fly. Could be fixed by caching Keras datasets locally in advance.

**Comment by cburgdorf** [comment]
Thank you @fchollet. We'll work on that to make that possible.

**Comment by cburgdorf** [comment]
Just wanted to give a quick status update here. We are currently working on a feature that lets users upload datasets to our platform which can then be mounted into the execution via the `ml.yaml` file. With this feature, we can put all the relevant datasets in our storage and configure the example labs so that they have the data mounted at runtime.

With this feature in place, anonymous users don't need internet access and we can make it so that they can run executions for a limited execution time. This should land within the next weeks.

**Comment by fchollet** [comment]
@cburgdorf is this still on the table?

**Comment by cburgdorf** [comment]
Hey François!

Sorry for the long silence here. We are actually still keen on getting this done. It's just that we had some other high priority issues to get done first. We [launched GPU support](https://blog.machinelabs.ai/2017/11/20/announcing-gpu-support-and-revisiting-the-neural-style-transfer/) in November and beginning this year [we replaced the editor](https://blog.machinelabs.ai/2018/01/08/machinelabs-is-now-in-public-beta/) with the same editor that Visual Studio Code uses. We also opened [public beta](https://blog.machinelabs.ai/2018/01/08/machinelabs-is-now-in-public-beta/) which means people don't need invites anymore to join the system. Everyone can now just login and directly run code.

However, anonymous users still can't test run executions yet, only look at existing, cached executions. We just haven't had the time to tackle this yet as we felt the other improvements had to come first. We are a team of just four developers (actually just ramped up to four this year!) so we haven't had the resources to get it done faster.

Having anonymous user being able to test run executions (for a couple of seconds and without internet access) is still your threshold here, correct?

**Comment by fchollet** [comment]
> Having anonymous user being able to test run executions (for a couple of seconds and without internet access) is still your threshold here, correct?

Yes. I still think this would be a nice feature to add to docs. At the same time, you have competitors (in particular Kaggle Kernels -- disclaimer: Kaggle is part of Google), and I will need to check anew whether your service is the best option to add this feature (with the guiding goal being to be maximally helpful to users reading the Keras docs).

**Comment by fchollet** [comment]
Closing PR, feel free to reopen if you have something. I still think this would be a very nice feature of the docs website.
